File: src/main/com/stockcomp/repository/InvestmentRepository.java
package com.stockcomp.repository;

import com.stockcomp.domain.contest.Investment;
import com.stockcomp.domain.contest.Participant;
import com.stockcomp.domain.contest.enums.ContestStatus;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface InvestmentRepository extends JpaRepository<Investment, Long> {

    Investment findBySymbolAndParticipant(String symbol, Participant participant);

    List<Investment> findAllByParticipant(Participant participant);

    @Query("SELECT i FROM Investment i join i.participant p join p.contest c where c.contestStatus  = ?1")
    List<Investment> findAllInvestmentsByContestStatus(ContestStatus contestStatus);
}

File: src/main/com/stockcomp/investment/repository/InvestmentRepository.kt
package com.stockcomp.investment.repository

import com.stockcomp.contest.entity.ContestStatus
import com.stockcomp.investment.entity.Investment
import com.stockcomp.participant.entity.Participant
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository

@Repository
interface InvestmentRepository : JpaRepository<Investment, Long> {

    fun findBySymbolAndParticipant(symbol: String, participant: Participant): Investment?

    fun findAllByParticipant(participant: Participant): List<Investment>

    @Query("SELECT i FROM Investment i join i.participant p join p.contest c where c.contestStatus  = ?1")
    fun findAllByContestStatus(contestStatus: ContestStatus): List<Investment>
}